### PR TITLE
Include silenced alerts in health popup

### PR DESCRIPTION
**Pull request by pcbailey**
## 📝 Description

Silenced alerts are included in the calculation for cluster health status that's displayed in the OCP dashboard. In order to calculate the correct number of alerts to display in the health popup, the silenced alerts needed to be included. The existing method of accessing alerts (`useDashboardResources`) doesn't include silenced alerts.

This PR:

- Adds hooks to retrieve silences and all alerts, including silenced ones
- Adds silenced alerts to the health popup
- Modifies the URL for the alerts links to include silenced alerts
- Modifies VirtualizationAlerts to avoid a type error*

* VirtualizationAlerts is part of code for the status card, which is no longer used and will be removed in a separate PR along with the inventory card.

### CODE REVIEW

File: src/utils/hooks/useAlerts/useAlerts.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+type UseAlerts = () => { alerts: Alert[]; loaded: boolean; loadError: unknown };
+
+const useAlerts: UseAlerts = () => {
+  const [response] = usePrometheusPoll({
+    endpoint: PrometheusEndpoint?.RULES,
+  });
+  const { silences } = useSilences();
+
+  const allAlerts = useMemo(() => {
+    const data = (response as PrometheusRulesResponse)?.data;
```
Comment (metalice) [inline src/utils/hooks/useAlerts/useAlerts.ts:24]: why we need as here? what usePrometheusPoll returns?

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    rule?.name,
+    rule?.duration,
+    rule?.query,
+    ...Object?.entries(rule?.labels).map(([k, v]) => `${v}=${k}`),
+  ].join(',');
+  return { ...rule, id: String(murmur3(key, MONITORING_SALT)) };
+};
+
+export const getAlertsAndRules = (
+  data: PrometheusRulesResponse['data'],
```
Comment (metalice) [inline src/utils/hooks/useAlerts/utils/utils.ts:32]: i think there is a data type for data

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+): { alerts: Alert[]; rules: Rule[] } => {
+  // Flatten the rules data to make it easier to work with, discard non-alerting rules since those
+  // are the only ones we will be using and add a unique ID to each rule.
+  const groups = data?.groups as PrometheusRulesResponse['data']['groups'];
```
Comment (metalice) [inline src/utils/hooks/useAlerts/utils/utils.ts:36]: there is a type for group at src/views/virtualmachines/details/tabs/overview/utils/utils.tsx

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  // Add rule object to each alert
+  const alerts = rules?.flatMap((rule) => rule?.alerts?.map((alert) => ({ rule, ...alert })));
+
+  return { alerts, rules };
+};
+
+export const isSilenced = (alert: Alert, silence: Silence): boolean =>
+  [AlertStates.Firing, AlertStates.Silenced].includes(alert.state) &&
+  silence?.matchers?.every((m) => {
```
Comment (metalice) [inline src/utils/hooks/useAlerts/utils/utils.ts:51]: Can you please add a more meaningful name for m :) ?

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    const isMatch = m.isRegex
+      ? new RegExp(`^${m.value}$`).test(alertValue)
+      : alertValue === m.value;
+    return m.isEqual === false && alertValue ? !isMatch : isMatch;
+  });
+
+export const silenceFiringAlerts = (alerts, silences): Alert[] =>
+  // For each firing alert, store a list of the Silences that are silencing it
+  // and set its state to show it is silenced
+  alerts?.map((alert) => {
```
Comment (metalice) [inline src/utils/hooks/useAlerts/utils/utils.ts:62]: you will get alerts[] with cells and empty values. filter isn't better here?

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    if (alert?.state !== AlertStates.Firing) {
+      return alert;
+    }
+
+    alert.silencedBy = silences?.filter(
```
Comment (metalice) [inline src/utils/hooks/useAlerts/utils/utils.ts:67]: u are running on silences each alert[x], can you run over it just once and send to isSilenced instead?
take silences.filter out of this scope and u have it ready

Comment (metalice) [reply src/utils/hooks/useAlerts/utils/utils.ts:67]: at leats the silence?.status?.state === SilenceStates.Active can be achieved by running over once

File: src/views/clusteroverview/OverviewTab/status-card/utils/VirtualizationAlerts.tsx [hunk 7]
```diff
@@ -1,18 +1,16 @@
 import * as React from 'react';
```
Comment (metalice) [inline src/views/clusteroverview/OverviewTab/status-card/utils/VirtualizationAlerts.tsx:1]: please remove *

File: src/utils/hooks/useAlerts/useAlerts.ts (hunk 1 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/useAlerts.ts:24]: `usePrometheusPoll` returns the `PrometheusResponse` type, whose `data` property has a value of `PrometheusData`. The data property of `PrometheusRulesResponse` has a value of `{ groups?: Group[] }`, which conforms to the shape of the data returned from the `PrometheusEndpoint.RULES` endpoint.

File: src/utils/hooks/useAlerts/utils/utils.ts (hunk 3 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts]: Fixed!

File: src/utils/hooks/useAlerts/utils/utils.ts (hunk 4 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts]: Good catch. I had to fix a lot of those. I should have known I was going to miss one. =)

File: src/utils/hooks/useAlerts/utils/utils.ts (hunk 6 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts]: I believe I did what you asked, but please double check.

File: src/utils/hooks/useAlerts/utils/utils.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  [AlertStates.Firing, AlertStates.Silenced].includes(alert.state) &&
+  silence?.matchers?.every((m) => {
+    const alertValue = alert?.labels?.[m.name] || '';
+    const isMatch = m.isRegex
+      ? new RegExp(`^${m.value}$`).test(alertValue)
+      : alertValue === m.value;
+    return m.isEqual === false && alertValue ? !isMatch : isMatch;
+  });
+
+export const silenceFiringAlerts = (alerts, silences): Alert[] =>
```
[the remark this answers is not in the document]
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts]: I broke it down some. I need to check with the original author on one part of it, but he's out of the office until tomorrow.

File: src/utils/hooks/useAlerts/utils/utils.ts (hunk 2 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts:23]: I didn't see it anywhere. If you can find it I'll be happy to use it.

Comment (metalice) [reply src/utils/hooks/useAlerts/utils/utils.ts:23]: can this help?
image

File: src/views/dashboard-extensions/KubevirtHealthPopup/KubevirtHealthPopup.tsx [hunk 9]
```diff
@@ -21,6 +21,7 @@ const KubevirtHealthPopup: FC = () => {
   const descriptionText = t(
     'You can host and manage virtualized workloads on the same platform as container-based workloads.',
   );
+  console.log('**** loaded: ', loaded);
```
Comment (metalice) [inline src/views/dashboard-extensions/KubevirtHealthPopup/KubevirtHealthPopup.tsx:24]: :)

File: src/utils/hooks/useAlerts/utils/utils.ts (hunk 2 above)
Comment (pcbailey) [reply src/utils/hooks/useAlerts/utils/utils.ts:23]: Unfortunately, no. This is what `PrometheusRulesResponse` looks like. The `data` field is different from `PrometheusResponse`. I don't know why they didn't create proper types for it on the console side.

```
export type PrometheusRulesResponse = {
  data: {
    groups?: Group[];
  };
  status: string;
};
```

File: src/utils/hooks/useKubevirtAlerts.ts [hunk 10]
```diff
@@ -1,27 +1,16 @@
 import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (pcbailey) [reply src/utils/hooks/useKubevirtAlerts.ts]: Fixed!

File: src/views/dashboard-extensions/KubevirtHealthPopup/KubevirtHealthPopup.tsx (hunk 9 above)
Comment (pcbailey) [reply src/views/dashboard-extensions/KubevirtHealthPopup/KubevirtHealthPopup.tsx]: Fixed!

**Pull request merged**